### ISSUE TITLE
[RFC]: Non-blocking core model loop

### ISSUE THREAD

**Issue by njhill** [vllm-project/vllm#43702]
# Motivation

The vLLM core engine process today **blocks** whenever it has no active requests. The non-DP loop ends up parked in `multiprocessing.Queue.get(block=True)`. In the DP case the wake-up is staged through a dedicated `DPCoordinator` process that broadcasts messages to bring all ranks out of their idle state together.

```python
# vllm/v1/engine/core.py - current non-DP busy loop
def run_busy_loop(self):
    while self._handle_shutdown():
        # 1) Poll the input queue until there is work to do.
        self._process_input_queue()        # blocks on input_queue.get(block=True)
        # 2) Step the engine core and return the outputs.
        self._process_engine_step()
```

The blocking model has worked well, but it has accumulated meaningful cost:

### 1. Multiplexing more than one wake-up source is awkward

The engine needs to react to more than just new requests: `DPCoordinator` wave messages, KV-connector completion / failure events, FSM/grammar compilation futures, and future signals like async weight updates and elastic-EP changes. Today these are multiplexed via a background ZMQ-poller thread that feeds an in-process `Queue` the main thread blocks on. It works, but every new signal source means another producer thread or another path into that queue.

### 2. Separate DP Coordinator process required

The `DPCoordinator` process is responsible for:

1. **Waking all DP ranks together when a new wave starts.** When a request lands at one idle rank, the coordinator broadcasts `START_DP_WAVE` so the ranks begin a new wave in lockstep. This is the dominant reason the coordinator exists.
2. **Maintaining wave numbering** across engines, including stale-wave handling for requests that arrive between coordinator broadcasts.
3. **Aggregating per-engine queue stats** and publishing them to the front-end(s) for the "internal DP LB" mode.
4. Handling elastic-EP add/remove events.

(1) and (2) only exist because the engine is allowed to fall asleep.

### 3. State machine complexity

Several distributed bits of state - `engines_running`, `current_wave`, `pending_pause`, plus the coordinator's mirror of them - all need to agree across processes. Most exist only to make "wake up together" work. In a non-blocking design "running" is no longer a state at all, just a per-step decision about whether to forward.

# Proposed Change

**Make the engine core loop non-blocking, and shrink the DP coordinator to its remaining intrinsic role.**

### 1. Core engine loop becomes a spin loop

The `EngineCoreProc.run_busy_loop` and `DPEngineCoreProc.run_busy_loop` paths converge on a single non-blocking shape:

```python
def run_busy_loop(self):
    while self._handle_shutdown():
        # Drain any pending client requests / control messages. Non-blocking.
        self._process_input_queue_nowait()

        # Drain any pending async signals (KV connector, grammar futures, etc.).
        self._process_async_signals_nowait()

        # Step the engine. In DP+EP mode this is a dummy batch when no local work
        # is ready but the per-iteration all-reduce says some peer rank has work.
        self._process_engine_step()
```

The background ZMQ-poller thread that feeds `input_queue` is kept; the main thread just stops sleeping on it.

The DP case no longer has a distinct "running" vs "stopped" state - every engine is always looping. What remains is a per-iteration **decision**: should this step run a model forward? In DP+EP that decision must agree across ranks. The dummy-batch / lockstep-forward contract itself is unchanged.

### Idle-throttling policy

A literal busy-spin would burn one CPU core per engine and per worker process even on a completely idle cluster (see Costs). The default policy:

* If there's work, iterate with no sleep. This is identical to today's "busy" DP+EP behaviour.
* If idle, `time.sleep(0)` or similar to yield the GIL, or a configurable short sleep (e.g. `VLLM_ENGINE_IDLE_SLEEP_MS`, default `0`).

The same lever applies to workers: when `dispatch_cg_and_sync_dp` (V2) or `coordinate_batch_across_dp` (V1, if adapted) returns the all-zero descriptor, the model runner sleeps for the configured interval before returning to the dequeue.

### 2. Subsuming the per-iteration all-reduce into the V2 model runner

The V2 model runner already does a per-step CPU-group all-reduce (`dispatch_cg_and_sync_dp`) that exchanges `(num_tokens, cg_mode, uniform_token_count)` across DP ranks and short-circuits when every rank reports zero tokens. This already answers "does any rank have work?" in the same collective that handles per-step DP padding.

In the V2 path we therefore drop the per-iteration outer all-reduce on `has_unfinished_requests` entirely. The loop becomes "always enter the runner; the runner's existing per-step sync handles both jobs." No `engines_running`, no `_has_global_unfinished_reqs`, no 32-step throttle.

The V1 model runner has a structurally identical per-step primitive (`coordinate_batch_across_dp`) but lacks the explicit all-zero short-circuit. Two options, deferrable:

* **Adapt V1 similarly** - add the short-circuit so V1 and V2 share the simpler loop.
* **Keep the outer all-reduce on the V1 path** until the V1 runner is removed, then delete the V1 branch whole. As a side benefit, V1 workers stay blocked on the SHM dequeue at idle rather than spinning.

### 3. Eliminate the DP wake-up machinery

With no engine ever blocking, the entire wave-broadcast path becomes dead code:

* The `START_DP_WAVE` request type and all engine-side wave bookkeeping.
* The `start_wave` engine->coordinator notification and corresponding broadcast.
* The XPUB/XSUB pub-sub leg between coordinator and engines.
* The `current_wave` field on `EngineCoreRequest`.

(Per section 2: under V2 the per-iteration cross-rank all-reduce also goes; under V1 it stays until the V1 runner is removed.)

### 4. Shrink (and eventually retire) the DP coordinator

After section 3 the only remaining responsibilities of `DPCoordinator` are stats aggregation for the "internal DP LB" mode (Python multi-api-server deployment need a near-real-time view of each engine's queue depth) and infrequent elastic-EP scaling events. It's function as a stats broadcaster is also redundant when using the new single-process rust front-end.

* **Short term:** keep `DPCoordinator` for internal-LB Python deployments, trimmed to just the stats-broadcast + control-plane role.

### 5. Simpler async-signal handling

With a spin loop, KV-connector events and grammar-compilation completions are just polled state each iteration. Grammar `Future.done()` becomes a normal step of the loop; future host-side KV signals (out-of-band cache hits, etc.) can be polled inline instead of plumbed through a wake-up queue.

## Benefits

* **Simpler engine state.** No `current_wave`, no `START_DP_WAVE`, no wake-up races, no stale-wave handling.
* **Simpler IPC topology.** The XPUB/XSUB coordinator<->engine fabric goes away.
* **Lower cold-request latency in DP deployments.** One fewer process and IPC hop on the request path.
* **Uniform code path for non-DP and DP.** The two `run_busy_loop` implementations converge.
* **Extension-friendly.** New async signal sources plug in as ordinary polled checks rather than as new wake-up producers.
* **Clear deprecation path** for the DP coordinator.

## Costs

* **CPU usage at idle.** Under V2 (and V1 + adapt option), every engine core *and* every GPU worker process spins at idle - roughly N + NxTxP cores for DP=N, TP=T, PP=P (e.g. ~12 cores for DP=4 TP=2, ~72 cores per node for DP=8 TP=8). Production deployments are rarely idle for long - GPU time is the binding cost - and `VLLM_ENGINE_IDLE_SLEEP_MS` applies to both engine cores and workers as the opt-in escape hatch for dev/test. (Under V1 + Option B in section 2, workers stay blocked on the SHM dequeue at idle as today; only the engine cores spin.)
* **Internal-LB Python deployments** continue to need the coordinator (but not if rust front-end is used.)

## Out of scope

* No change to KV-connector contracts, structured-output backends, or the scheduler itself.

### What other serving engines do

* **SGLang** - scheduler loop busy-spins by default; opt-in `--sleep-on-idle` switches to `zmq.Poller.poll(1000)`. No cross-rank wake signal, so `--sleep-on-idle` is effectively incompatible with DP+EP (idle ranks can stall peers in the per-step collective for up to 1 s).
* **TokenSpeed** - unconditional spin; no blocking path. DP coordination is a per-step CPU-group `all_gather` of rank state, with idle ranks running a dummy forward.
* **TensorRT-LLM** - block-waits at idle, but only on global rank 0. Request intake is centralized on rank 0; every step runs a global broadcast rooted at rank 0, and non-zero ranks sleep inside that collective. No coordinator process; the per-step broadcast is the wake-up.

### Any Other Things.

Claude was used to help draft this RFC, with significant human editing/rework.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by noooop** [comment]
I’m not sure if switching to an NCCL-based broadcast would be faster than ZMQ, because NCCL uses NVLink, NVSwitch, and InfiniBand, allowing it to bypass the Linux kernel during communication, whereas a ZMQ poller requires kernel involvement.

````
if rank == 0:
    tensor = torch.tensor([42.0], device=device)
else:
    tensor = torch.tensor([0.0], device=device)

dist.broadcast(tensor, src=0)

.....

dist.barrier()
torch.cuda.synchronize(device)
````

Moreover, we have already executed init_process_group for each worker, so we can use NCCL for free.

***

It would be perfect if we could epoll CUDA events stream fd whatever.


***

The content below is AI-generated; Hope it will provide some inspiration.



```python
import os
import select
import ctypes
import torch

# Get CUDA runtime library
libcudart = ctypes.CDLL("libcudart.so")

# Define the function signature of cudaLaunchHostFunc
libcudart.cudaLaunchHostFunc.restype = ctypes.c_int
libcudart.cudaLaunchHostFunc.argtypes = [
    ctypes.c_void_p,                # cudaStream_t
    ctypes.CFUNCTYPE(None, ctypes.c_void_p),  # callback function
    ctypes.c_void_p                 # user data
]

# Create an eventfd (non-blocking)
efd = os.eventfd(0, os.EFD_NONBLOCK)

# Callback function: write 1 to eventfd
# Note: This callback will be executed on a CUDA internal thread, it's safe to call Python functions
# (During epoll_wait, the main thread releases the GIL, so no deadlock)
def stream_done_callback(user_data):
    os.write(efd, b'\x01')

# Wrap as C callback type
CALLBACK = ctypes.CFUNCTYPE(None, ctypes.c_void_p)
callback = CALLBACK(stream_done_callback)

# Prepare a PyTorch stream and get the underlying cudaStream_t
stream = torch.cuda.Stream()
# PyTorch 1.9+ provides .cuda_stream attribute
cuda_stream_ptr = stream.cuda_stream
# For older versions, you can use stream._as_parameter_ or ctypes

# Submit some work to the stream (example: a dummy kernel to simulate)
with torch.cuda.stream(stream):
    a = torch.randn(1000, device='cuda')
    b = torch.randn(1000, device='cuda')
    c = a + b

# Register a callback on the stream to write to eventfd upon completion
res = libcudart.cudaLaunchHostFunc(
    ctypes.c_void_p(cuda_stream_ptr),
    callback,
    None
)
assert res == 0, f"cudaLaunchHostFunc failed: {res}"

# ----- Use epoll to wait for stream completion -----
epoll = select.epoll()
epoll.register(efd, select.EPOLLIN)

events = epoll.poll()  # Will block until the stream completes and writes to eventfd
for fd, event in events:
    if fd == efd and event & select.EPOLLIN:
        # Read data to clear the state (must be 8 bytes)
        os.read(efd, 8)
        print("CUDA stream done！")

# Cleanup
epoll.unregister(efd)
epoll.close()
os.close(efd)
```

**Comment by njhill** [comment]
Thanks @noooop, actually this proposal is kind of orthogonal to the use of zmq. I think we'd still use zmq for the input/output queues between the front-end process and the engine(s).

**Comment by noooop** [comment]
I see. The NCCL-based solution can only be used between runners. not for frontend.

There is at least one method to simultaneously and non-blockingly epoll ZMQ and CUDA stream done, but indeed, I'm not sure what use this solution has.
